### PR TITLE
Fix Issue 23176 - teach -vasm to print immediates properly.

### PR DESCRIPTION
**Pull request by maxhaton**
This also makes the hopefully non-controversial change of
introducing a nested function rather than a new use of goto.

### CODE REVIEW

File: src/dmd/backend/disasm86.d [hunk 1]
```diff
@@ -3621,7 +3659,16 @@ unittest
         "BF 00 00 00 00      mov  EDI,0",
         "41 0F C7 09         cmpxchg8b [R9]",
         "49 0F C7 09         cmpxchg16b [R9]",
-        "0F 01 F9            rdtscp"
+        "0F 01 F9            rdtscp",
+        "66 41 0F 70 C7 66        pshufd    XMM0,XMM15,066h",
```
Comment (ljmf00) [inline src/dmd/backend/disasm86.d:3663]: Whitespace alignment is not consistent with the previous ones.

File: src/dmd/backend/disasm86.d [hunk 2]
```diff
@@ -3621,7 +3659,16 @@ unittest
...
+        "F2 41 0F 70 C7 F1        pshuflw   XMM0,XMM15,0F1h",
+        "F3 41 0F 70 C7 C2        pshufhw   XMM0,XMM15,0C2h",
+        "66 41 0F C6 C7 CF        shufpd    XMM0,XMM15,0CFh",
+        "66 0F C6 00 CF           shufpd    XMM0,[RAX],0CFh",
+        "66 0F C2 00 CF           cmppd     XMM0,[RAX],0CFh",
+        "F3 41 0F C2 C7 AF        cmpss     XMM0,XMM15,0C7h",
+        "66 0F 73 FF 99           pslldq    XMM7,099h",
+
```
Comment (ljmf00) [inline src/dmd/backend/disasm86.d:3671]: Remove extra newline

File: src/dmd/backend/disasm86.d [hunk 3]
```diff
@@ -1776,7 +1776,27 @@ void puts(const(char)* s)
...
+mixin template PatternFunctions()
+{
+    /**
+     * Implements `xmm1, xmm2/m128, imm8`
+     * Params:
+     *   indexOffset = this will be called from various amounts of
+     *                 lookahead into the code buffer which means an index is needed
+     *                 for finding the right offset to the register.
+     */
+    void xmm_xmm_imm8(uint indexOffset)
```
Comment (ljmf00) [inline src/dmd/backend/disasm86.d:1793]: Why this needs to be in a mixin? Only used one time and restrictive for this use case (where parameters are in the parent context scope).

Comment (maxhaton) [reply src/dmd/backend/disasm86.d:1793]: Because there will be more than one of them

Comment (maxhaton) [reply src/dmd/backend/disasm86.d:1793]: And look at how messy the place it is mixin-ed is

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**